### ISSUE TITLE
Memory not released after deleting a layer, "Convert to Labels", "Convert to Layer" or "Convert datatype"

### ISSUE THREAD

**Issue by Karol-G** [napari/napari#4485]
## 🐛 Bug

Napari does not release old memory if one of the following actions is executed:
- The layer is deleted (Layer is still in memory)
- An image layer is converted into a labels layer (Image layer is still in memory)
- A labels layer is converted into an image layer (Labels layer is still in memory)
- The data type of a labels layer is changed (Labels layer with old data type is still in memory)

## To Reproduce

Example for deleting a layer:
1. Start napari
2. Open the python console
3. `import numpy as np`
4. `viewer.add_image(np.zeros((1000, 1000, 1000)), name="test")`
5. Observe the increase of memory consumption
6. Delete the layer
7. Observe that there is no decrease in memory consumption

The other issues can be reproduced similarly.

## Expected behavior

Memory should be released after any of those actions.

## Environment

```
napari: 0.4.15
Platform: Linux-5.13.0-40-generic-x86_64-with-glibc2.17
System: Ubuntu 20.04.3 LTS
Python: 3.8.12 (default, Oct 12 2021, 13:49:34) [GCC 7.5.0]
Qt: 5.15.2
PyQt5: 5.15.6
NumPy: 1.22.3
SciPy: 1.8.0
Dask: 2021.12.0
VisPy: 0.10.0

OpenGL:
- GL version: 4.6.0 NVIDIA 470.103.01
- MAX_TEXTURE_SIZE: 32768

Screens:
- screen 1: resolution 2560x1440, scale 1.0

Plugins:
- PartSeg Analysis project reader: 0.13.15
- PartSeg Image reader: 0.13.15
- PartSeg Image tiff writer: 0.13.15
- PartSeg Mask project reader: 0.13.15
- PartSeg Mask project writer: 0.13.15
- PartSeg Masked Tiff reader: 0.13.15
- PartSeg widgets: 0.13.15
- console: 0.0.4
- napari-accelerated-pixel-and-object-classification: 0.6.5
- napari-medical-image-formats-read-write: 0.4.2
- napari-time-slicer: 0.4.3
- napari-tools-menu: 0.1.10
- scikit-image: 0.4.15
- svg: 0.1.5
```

Best,
Karol

**Comment by Czaki** [comment]
Did you call `gc.collect()` from `gc` module to be sure that python free all unused memory?

**Comment by Karol-G** [comment]
No, I did not. But this are just steps to easily reproduce the error. This also happens without the python console a.k.a loading and converting data via the gui.
In my case, it often happens that I load a large image and its segmentation, then adapt the datatypes, which then leads to napari running out of memory.

**Comment by Czaki** [comment]
Hm. I could confirm that problem exists, but when reproduce using console it is caused by keeping reference by ipython console: 

![Zrzut ekranu z 2022-05-08 20-14-32](https://user-images.githubusercontent.com/3826210/167309676-f04ee2d4-0fdb-4d44-bc1c-76515f97715b.png)

Could you describe exactly steps that you perform? Exactly which plugins you use? It may simpler to find what keep reference to numpy arrays.

**Comment by Karol-G** [comment]
I have some plugins installed, but they are not used in this case. All I do is to load a large 3D image and convert it with one of the native napari operations specified in my first post. I am pretty sure that all data (small images, large images, 2D/3D) are affected by this, but one only notices it with large images.

I would suggest to generate a large 3D .tif image, save it to disk, load it into napari and convert the data type. This is probably the easiest method for you to reproduce this bug without the ipython console.

`pip install tifffile`

```python
import tifffile
import numpy as np

data = np.ones((1000, 1000, 1000), dtype=np.int32)
tifffile.imwrite('temp.tif', data)
```

**Comment by Czaki** [comment]
When trying to reproduce using `Convert to Image` and `Convert to Labels` I see increase of memory, but the call of `gc.collect()` free memory. 
Also, trying to play with weakref to memory shows that the NumPy array was released.

**Comment by Karol-G** [comment]
But I don't think that helps a user who soley works with the napari gui :/

**Comment by Czaki** [comment]
> But I don't think that helps a user who soley works with the napari gui :/

but if you could check if in your case `gc.collect()` (imported using `import gc`) reduce memory usage. If yes, then one of the napari maintainers could force `collect` call in multiple places.

**Comment by Karol-G** [comment]
I tested it now, but `gc.collect()` did not free any memory in my case.

**Comment by Czaki** [comment]
At this moment I have no more idea how to debug this. If I have more time I will write documentation about debugging memory leaks. But maybe someone else will have an idea of how to debug this.

**Comment by psobolewskiPhD** [comment]
I've also run into this with 0.4.18 and current main napari 514b40ae19b8c85c0f546976dbcccb29c505095d
See: https://napari.zulipchat.com/#narrow/stream/212875-general/topic/GL_MAX_TEXTURE.20sizes/near/404934321
When toggling to 3D and then deleting the layer, gc.collect() does free some of the memory -- that which was used for 3D display, but not all. deleting the data array (`del data`) also does not free the memory.
